### PR TITLE
refactor http abstraction

### LINKED ISSUE #301: [doc] client setup with httplug
the doc is still talking about guzzle, update for httplug
http://foshttpcache.readthedocs.io/en/latest/proxy-clients.html

### LINKED ISSUE #305: how to HttpAdapter?
we currently have [HttpAdapter](https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/ProxyClient/Http/HttpAdapter.php) that creates requests and sends them through a Httplug client. could a Httplug plugin duplicate requests for each caching proxy or is that not something that not feasable? any other solutions or architecture ideas?

probably we should at least not instantiate the adapter in the AbstractProxyClient https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/ProxyClient/AbstractProxyClient.php#L85

### PR DESCRIPTION
**Pull request by dbu**
fix #305, fix #301

the proxy clients know what they want to invalidate. the logic how to send the requests to each of the caching reverse proxy servers is now isolated in HttpAdapter, along with queuing the requests and sending them out in parallel on flush.

* [x] Clean up tests
* [x] Update documentation
* [x] Convert HTTP errors to exceptions in client

### CODE REVIEW

File: tests/Unit/ProxyClient/VarnishTest.php [hunk 1]
```diff
@@ -11,22 +11,40 @@
...
+        $varnish = new Varnish($this->client);
+        $this->client->expects($this->once())
+            ->method('invalidate')
+        ;
+        $this->client->expects($this->once())
+            ->method('flush')
+        ;
         $varnish->ban([])->flush();
 
+        /*
```
Comment (dbu) [inline tests/Unit/ProxyClient/VarnishTest.php]: clean up this class

**Comment by dbu** [comment]
thanks for the input. you are right, i will refactor to use the RequestFactory instead.

but the overall architecture seems right to you?

File: src/ProxyClient/Http/HttpAdapter.php [hunk 2]
```diff
@@ -254,6 +270,13 @@ private function setBaseUri($uriString = null)
      */
     private function filterUri($uriString, array $allowedParts = [])
     {
+        if (!is_string($uriString)) {
+            throw new \InvalidArgumentException(sprintf(
+                'Uri parameter must be a string, %s given',
```
Comment (sagikazarmark) [inline src/ProxyClient/Http/HttpAdapter.php]: Just for consistency: use URI or Uri

File: src/ProxyClient/Http/HttpAdapter.php [hunk 3]
```diff
@@ -254,6 +270,13 @@ private function setBaseUri($uriString = null)
...
+                gettype($uriString)
```
Comment (sagikazarmark) [inline src/ProxyClient/Http/HttpAdapter.php]: Maybe output object instance and/or resource types too? I saw a oneliner for that somewhere

**Comment by sagikazarmark** [comment]
Found a few more things.

I like the architecture. There is one more things which is not critical, but this is the time to think about it once more: I would try to find a better name for HttpAdapter, because it's not really an adapter. Maybe HttpBridge?

**Comment by dbu** [comment]
naming things... what about calling it HttpInvalidator?

**Comment by sagikazarmark** [comment]
Does it do any invalidation? As far as I saw it's a bridge between HTTP Client and ProxyClient...hence the Bridge idea. :smile: 

If it has any more behaviour, it's fine too.

**Comment by dbu** [comment]
it accepts PSR 7 requests, queues them and then sends them to a list of servers. while sending them out, it also completes them with host header and sends them to the right ips.

the requests are meant to be invalidation requests, but could be other things. so i am not sure if bridge is the right word. the fact that this is HTTP is also known in the proxy client implementations.

File: src/ProxyClient/Http/HttpAdapter.php (hunk 3 above)
Comment (dbu) [inline src/ProxyClient/Http/HttpAdapter.php]: don't think it will help much, when it has to be a string and nothing else

**Comment by sagikazarmark** [comment]
Hm.... HttpDispatcher, HttpProxynator, HttpInvalidationBroker, HttpConnector?

HttpInvalidator is not actually true then as it doesn't do invalidation, it's the server that does it.

**Comment by dbu** [comment]
dispatcher sounds good! thanks for the help mark.

**Comment by sagikazarmark** [comment]
Anytime, although Proxynator was my personal favourite. :smile:

**Comment by dbu** [comment]
> Proxynator 

maybe we should rename the library to FOSHttpProxynator :-)

**Pull request merged**